File: src/Year2017/Day23.php
<?php

namespace Mzentrale\AdventOfCode\Year2017;

use Mzentrale\AdventOfCode\Puzzle;

class Day23 implements Puzzle
{
    public function part1(string $input)
    {
        $instructions = $this->parseInput($input);
        $registries   = array_combine(range('a', 'h'), array_fill(0, 8, 0));
        return $this->runWithRegistries($instructions, $registries);
    }

    public function part2(string $input)
    {
        for ($b = 109300, $h = 0; $b <= 126300; $b += 17) {
            if (!$this->isPrime($b)) $h += 1;
        }
        return $h;
    }

    private function isPrime(int $num): bool
    {
        if (($num > 2 && $num % 2 === 0) || $num === 1) return false;
        for ($i = 3, $root = sqrt($num) + 1; $i < $root; $i += 2) {
            if ($num % $i === 0) return false;
        }
        return true;
    }

    private function parseInput(string $input): array
    {
        return array_map(function (string $row) {
            return explode(' ', $row);
        }, explode("\n", trim($input)));
    }

    private function runWithRegistries(array $instructions, array $registries): int
    {
        $value = function (string $what) use (&$registries): int {
            return $registries[$what] ?? $what;
        };

        for ($i = 0, $num = count($instructions), $mul = 0; $i < $num; $i++) {
            list($operation, $which, $qty) = $instructions[$i];
            switch ($operation) {
                case 'set':
                    $registries[$which] = $value($qty);
                    break;
                case 'sub':
                    $registries[$which] -= $value($qty);
                    break;
                case 'mul':
                    $registries[$which] *= $value($qty);
                    $mul++;
                    break;
                case 'jnz':
                    if ($value($which) !== 0) {
                        $i += $value($qty) - 1;
                    }
                    break;
            }
        }

        return $mul;
    }
}

File: python/year2017/day23.py
from math import sqrt

def is_prime(num):
    if (num > 2 and num % 2 == 0) or num == 1:
        return False
    i, max = 3, sqrt(num)
    while i <= max:
        if num % i == 0:
            return False
        i += 2
    return True

def simplify():
    b, c, h = 109300, 126300, 0
    for n in range(b, c + 1, 17):
        if not(is_prime(n)):
            h += 1
    return h

def run_instructions(a=0):
    b = c = d = e = f = g = h = 0

    b = c = 93
    if a:
        b = 109300
        c = 126300

    while True:
        f = 1
        d = 2
        while True:
            e = 2
            while True:
                g = d * e - b
                if not(g):
                    f = 0
                e += 1
                g = e - b
                if not(g):
                    break
            d += 1
            g = d - b
            if not(g):
                break
        if not(f):
            h += 1
        g = b - c
        if not(g):
            break
        b += 17

    return h

print simplify()
print run_instructions(1)
